### PR TITLE
refactor GroupBy internals

### PR DESCRIPTION
**Pull request by dcherian**
1. Simplify ResolvedGrouper by moving logic to EncodedGroups
2. Stack outside ResolvedGrouper in GroupBy.__init__ to prepare for
   multi-variable GroupBy

Broken out form #9372 

- [x] Tests added

### CODE REVIEW

File: xarray/core/groupby.py [hunk 1]
```diff
@@ -1362,17 +1331,14 @@ def apply(self, func, args=(), shortcut=None, **kwargs):
...
         """Recombine the applied objects like the original."""
         applied_example, applied = peek_at(applied)
-        coord, dim, positions = self._infer_concat_args(applied_example)
+        dim, positions = self._infer_concat_args(applied_example)
         combined = concat(applied, dim)
-        (grouper,) = self.groupers
-        combined = _maybe_reorder(combined, dim, positions, N=grouper.group.size)
+        combined = _maybe_reorder(combined, dim, positions, N=self.group1d.size)
         # assign coord when the applied function does not return that coord
-        if coord is not None and dim not in applied_example.dims:
```
Comment (dcherian) [inline xarray/core/groupby.py:1370]: nice simplification!

File: xarray/groupers.py [hunk 2]
```diff
@@ -57,18 +69,49 @@ class EncodedGroups:
...
+
+    def __init__(
+        self,
+        codes: DataArray,
+        full_index: pd.Index,
+        group_indices: GroupIndices | None = None,
+        unique_coord: Variable | _DummyGroup | None = None,
+        coords: Coordinates | None = None,
+    ):
+        from xarray.core.groupby import _codes_to_group_indices
```
Comment (dcherian) [inline xarray/groupers.py:84]: logic moved from resolvedgrouper

File: xarray/core/groupby.py [hunk 3]
```diff
@@ -808,15 +773,27 @@ def _flox_reduce(
 
         # we did end up reducing over dimension(s) that are
         # in the grouped variable
-        group_dims = grouper.group.dims
-        if set(group_dims).issubset(set(parsed_dim)):
-            result = result.assign_coords(
-                Coordinates(
-                    coords={name: (name, np.array(output_index))},
-                    indexes={name: PandasIndex(output_index, dim=name)},
+        group_dims = set(grouper.group.dims)
```
Comment (dcherian) [inline xarray/core/groupby.py:776]: This is the only main logic change. It is needed to get  the new tests to pass

**Review by max-sixty — APPROVED** [review]
LGTM (was a quick scan, can look more tomorrow if helpful)

**Comment by dcherian** [comment]
Thanks! That's good enough. It mostly undoes a previous refactor, haha.

**Pull request merged**